functions/go/set-namespace/transformer/namespace.go

```diff
@@ -100,68 +101,47 @@ func VisitNamespaces(objects []*fn.KubeObject, visitor func(namespace *Namespace
 		case o.IsLocalConfig():
 			continue
 		case o.IsGVK("v1", "Namespace"):
-			namespace := o.GetStringOrDie("metadata", "name")
+			namespace := o.NestedStringOrDie("metadata", "name")
 			visitor(NewNamespace(o, &namespace))
 			o.SetOrDie(&namespace, "metadata", "name")
 		case o.IsGVK("apiextensions.k8s.io/v1", "CustomResourceDefinition"):
-			namespace := o.GetStringOrDie("spec", "conversion", "webhook", "clientConfig", "service", "namespace")
+			namespace := o.NestedStringOrDie("spec", "conversion", "webhook", "clientConfig", "service", "namespace")
 			visitor(NewNamespace(o, &namespace))
 			o.SetOrDie(&namespace, "spec", "conversion", "webhook", "clientConfig", "service", "namespace")
 		case o.IsGVK("apiregistration.k8s.io/v1", "APIService"):
-			namespace := o.GetStringOrDie("spec", "service", "namespace")
+			namespace := o.NestedStringOrDie("spec", "service", "namespace")
 			visitor(NewNamespace(o, &namespace))
 			o.SetOrDie(&namespace, "spec", "service", "namespace")
 		case o.GetKind() == "ClusterRoleBinding" || o.GetKind() == "RoleBinding":
 			subjects := o.GetSlice("subjects")
 			for _, s := range subjects {
+				// Note: Users may not expect to change the subject namespace fields. We shall provide an option
+				// to skip changing the namespace in ClusterRolebinding/Rolebinding subjects.
+				// See https://github.com/kubernetes-sigs/kustomize/issues/1599
 				var ns string
 				found, _ := s.Get(&ns, "namespace")
 				if found {
 					visitor(NewNamespace(o, &ns))
-					_ = s.Set(&ns, "namespace")
+					_ = s.SetNestedField(&ns, "namespace")
 				}
 			}
 			o.SetOrDie(&subjects, "subjects")
-		default:
-			// skip the cluster scoped resource. We determine if a resource is cluster scoped by
-			// checking if the metadata.namespace is configured.
 		}
-		if o.HasNamespace() {
-			// We made the hypothesis that a namespace scoped resource should have its metadata.namespace field setup.
-			// All above special types are cluster scoped, except RoleBinding
-			namespace := o.GetStringOrDie("metadata", "namespace")
+		// "IsNamespaceScoped" detects resource scope by checking whether `metadata.namespace` is set.
+		// This is valid hypothesis is for custom k8s resources. For k8s resources, the function always gives accurate scope information.
+		if o.IsNamespaceScoped() {
```

Comment:
Looking at the behavior, IMO `HasNamespace` is a better name than `IsNamespaceScoped`.

Reply:
HasNamespace and IsNamespaceScoped are actually different.
`HasNamespace` only checks whether the `metadata.namespace` is set, this is our hypothesis to tell whether a custom KRM resources is namespace scoped or not.

`IsNamespaceScoped` is the name from "func IsXXX() bool" and k8s CRD "scope: namespaced", which expects to tell more accurate namespace info, which is aligned with what the SDK does:  
The SDK now has a better knowledge for core k8s resource like `Deployment`, `Service`  and `RoleBinding`, and does not need to check the `metadata.namespace`  https://github.com/GoogleContainerTools/kpt-functions-sdk/pull/568.

Reply:
I misunderstood something :P